rbac/providers/ldap/ldap_payload_transformer.py

```diff
@@ -0,0 +1,68 @@
+#! /usr/bin/env python3
+
+# Copyright 2018 Contributors to Hyperledger Sawtooth
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+from datetime import datetime
+
+
+# These functions are strictly for transforming and mapping.
+# There should be no throwing of exceptions in this file. If an exception does need to
+# be thrown, it should be thrown by the ldap_message_validator before this conversion takes place.
+
+
+# TODO: Write a unit test for this. Add to the Docstring an actual example.
+def to_date_ldap_query(rethink_timestamp):
+    """
+        Call to transform timestamp stored in RethinkDB to a string in the following format:YYYYmmddHHMMSS.Tz
+    """
+    return datetime.strptime(
+        rethink_timestamp.split("+")[0], "%Y-%m-%dT%H:%M:%S.%f"
+    ).strftime("%Y%m%d%H%M%S.0Z")
+
+
+def to_user_create(next_payload):
+    """Converts a NEXT payload to an ldap payload"""
```

Comment:
I saw the filter, and it was breaking for me in mysterious ways because:

  - It takes 2 lists of attributes, sorts them, proceeds to map while also doing validation (ie the job of a validator). If an attribute is missing, it fails with the generic Python missing attribute error. Thus, I have separated (in the ldap directory) validation from transformation to make errors easier to troubleshoot.

  I mentioned a few times this sprint that moving functionality into a 'common' directory was never agreed upon, and that I believe it was carried out before the features themselves were fully implemented. At the same time: I have committed to deliver AD sync. The fastest way for me to do that without breaking AAD sync is by isolating the functionality to 'ldap' until the feature is stable.

I have objections to interleaving AD and AAD mappings to NEXT. Architecturally, mixing it all together does not support adding new providers in a straightforward way. (ie you end up at a loop on error pointing to mappings of { key: NEXT value {'ldap':'foo', 'azure':'bar'}} , etc. It complicates the process of determining which part of the bind was missing what field and how to correct it.

Reply:
Makes sense.